delta-lake/src/main/java/org/apache/iceberg/BaseMigrateDeltaLakeTableAction.java

```diff
@@ -0,0 +1,322 @@
+    Schema schema = convertDeltaLakeSchema(updatedSnapshot.getMetadata().getSchema());
+    PartitionSpec partitionSpec = getPartitionSpecFromDeltaSnapshot(schema);
+    // TODO: check whether we need more info when initializing the table
+    Table icebergTable =
+        this.icebergCatalog.createTable(
+            newTableIdentifier,
+            schema,
+            partitionSpec,
+            destTableProperties(
+                updatedSnapshot, this.deltaTableLocation, this.additionalProperties));
+
+    copyFromDeltaLakeToIceberg(icebergTable, partitionSpec);
+
+    Snapshot snapshot = icebergTable.currentSnapshot();
+    long totalDataFiles =
+        Long.parseLong(snapshot.summary().get(SnapshotSummary.TOTAL_DATA_FILES_PROP));
+    LOG.info(
+        "Successfully loaded Iceberg metadata for {} files to {}",
+        totalDataFiles,
+        deltaTableLocation);
+    return new BaseMigrateDeltaLakeTableActionResult(totalDataFiles);
+  }
+
+  /** TODO: check the correctness for nested schema */
+  private Schema convertDeltaLakeSchema(io.delta.standalone.types.StructType deltaSchema) {
+    Type converted =
+        DeltaLakeDataTypeVisitor.visit(deltaSchema, new DeltaLakeTypeToType(deltaSchema));
+    return new Schema(converted.asNestedType().asStructType().fields());
+  }
+
+  private PartitionSpec getPartitionSpecFromDeltaSnapshot(Schema schema) {
+    List<String> partitionNames = deltaLog.snapshot().getMetadata().getPartitionColumns();
+    if (partitionNames.isEmpty()) {
+      return PartitionSpec.unpartitioned();
+    }
+
+    PartitionSpec.Builder builder = PartitionSpec.builderFor(schema);
+    for (String partitionName : partitionNames) {
+      builder.identity(partitionName);
+    }
+    return builder.build();
+  }
+
+  private void copyFromDeltaLakeToIceberg(Table table, PartitionSpec spec) {
+    // TODO: double check the arguments' meaning here
+    Iterator<VersionLog> it =
+        deltaLog.getChanges(
+            0, // retrieve actions starting from the initial version
+            false); // not throw exception when data loss detected
+
+    while (it.hasNext()) {
+      VersionLog versionLog = it.next();
+      List<Action> actions = versionLog.getActions();
+
+      // We first need to iterate through to see what kind of transaction this was. There are 3
+      // cases:
+      // 1. AppendFile - when there are only AddFile instances (an INSERT on the table)
+      // 2. DeleteFiles - when there are only RemoveFile instances (a DELETE where all the records
+      // of file(s) were removed
+      // 3. OverwriteFiles - when there are a mix of AddFile and RemoveFile (a DELETE/UPDATE)
+
+      // Create a map of Delta Lake Action (AddFile, RemoveFile, etc.) --> List<Action>
+      Map<String, List<Action>> deltaLakeActionMap =
+          actions.stream()
+              .filter(action -> action instanceof AddFile || action instanceof RemoveFile)
```

Comment:
why filter here?

Reply:
DeltaLog's action has the following action types:  [AddCDCFile](https://delta-io.github.io/connectors/latest/delta-standalone/api/java/io/delta/standalone/actions/AddCDCFile.html "class in io.delta.standalone.actions"), [AddFile](https://delta-io.github.io/connectors/latest/delta-standalone/api/java/io/delta/standalone/actions/AddFile.html "class in io.delta.standalone.actions"), [CommitInfo](https://delta-io.github.io/connectors/latest/delta-standalone/api/java/io/delta/standalone/actions/CommitInfo.html "class in io.delta.standalone.actions"), [Metadata](https://delta-io.github.io/connectors/latest/delta-standalone/api/java/io/delta/standalone/actions/Metadata.html "class in io.delta.standalone.actions"), [Protocol](https://delta-io.github.io/connectors/latest/delta-standalone/api/java/io/delta/standalone/actions/Protocol.html "class in io.delta.standalone.actions"), [RemoveFile](https://delta-io.github.io/connectors/latest/delta-standalone/api/java/io/delta/standalone/actions/RemoveFile.html "class in io.delta.standalone.actions"), [SetTransaction](https://delta-io.github.io/connectors/latest/delta-standalone/api/java/io/delta/standalone/actions/SetTransaction.html "class in io.delta.standalone.actions")

Among them, `AddCDCFile`, `AddFile`, and `RemoveFile` are related to the change of data files. Hence, we may want to filter out other action types here.

For `AddCDCFile`, based on my understanding, the CDC support for iceberg is still under development:(#3941 ). Also, I am not sure if we need to include the CDC data during the migration process. Hence, currently I choose to preserver the previous implementation which handles only the `AddFile` and `RemoveFile`.